### ISSUE TITLE
Support custom stan models

### ISSUE THREAD

**Issue by jlindenger** [facebook/prophet#340]
We're using the prophet library with some of our own custom stan models, but the way that the current "linear" and "logistic" models are loaded on import and managed in the local scope of the Prophet.fit method makes this rather challenging.

I propose making model a property on the Prophet class that gets set appropriately to "linear", "logistic", or custom during __init__, and tweaking models.py to load the pickled stan models lazily upon request to avoid the memory impact of loading them on import.

I'm happy to put together a PR for this. We've already done most of it to make our own code work.

**Comment by bletham** [comment]
This is a great idea, I'd love to see a more convenient interface for custom models.

The main challenge that I see is handling the additional model-specific code that is not in Stan. In particular,
* Initialization for the optimization ([linear](https://github.com/facebook/prophet/blob/master/python/fbprophet/forecaster.py#L677) and [logistic](https://github.com/facebook/prophet/blob/master/python/fbprophet/forecaster.py#L701))
* Evaluation of the trend model ([linear](https://github.com/facebook/prophet/blob/master/python/fbprophet/forecaster.py#L888) and [logistic](https://github.com/facebook/prophet/blob/master/python/fbprophet/forecaster.py#L915)).

Perhaps the best thing to do will be to make the model a class that would include these two methods, and the stan code / compiled pickle. The complicated part is that these functions already have slightly different inputs (logistic expects `cap`, linear does not), and the parameters k, m, delta, beta, and sigma_obs are hard-coded throughout. It would be a bit of a refactor to support Stan code that has different parameters. That said, I think even requiring these same parameters would still allow flexibility in things like the likelihood model (#337) and so would be useful. Also see #264 for adding some new models, and the type of things that I would hope would be cleaner with this. I'd be curious to hear about what sort of model changes you have found useful.

**Comment by bletham** [comment]
I think #501 will make this much easier by making the Stan code the only place that you have to worry about modifying the model, we can revisit this once that is finished.

**Comment by bletham** [comment]
In light of #501 not being feasible, I don't expect this will be either. I did add to the documentation that PR #1466 provides a good template for what needs to be done to change the trend (https://facebook.github.io/prophet/docs/additional_topics.html#flat-trend-and-custom-trends) . For now I think the best course will be to leave it to the user to follow that PR if a different type of trend function is desired.
